service_container/service_locators.rst

```diff
@@ -0,0 +1,227 @@
+                </service>
+
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        use Symfony\Component\DependencyInjection\ServiceLocator;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        //...
+
+        $container
+            ->register('app.command_handler_locator', ServiceLocator::class)
+            ->addTag('container.service_locator')
+            ->setArguments(array(
+                'AppBundle\FooCommand' => new Reference('app.command_handler.foo'),
+                'AppBundle\BarCommand' => new Reference('app.command_handler.bar'),
+            ))
+        ;
+
+.. note::
+
+    The services defined in the service locator argument must be keyed.
+    Those keys become their unique identifier inside the locator.
+
+
+Now you can use it in your services by injecting it as needed:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        services:
+
+            AppBundle\CommandBus:
+                arguments: ['@app.command_handler_locator']
+
+    .. code-block:: xml
+
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+
+                <service id="AppBundle\CommandBus">
+                    <argument type="service" id="app.command_handler.locator" />
+                </service>
+
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        use AppBundle\CommandBus;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        //...
+
+        $container
+            ->register(CommandBus::class)
+            ->setArguments(array(new Reference('app.command_handler_locator')))
+        ;
+
+.. tip::
+
+    You should create and inject the service locator as an anonymous service if
+    it is not intended to be used by multiple services
+
+Usage
+-----
+
+Back to our CommandBus which now looks like::
+
+    // ...
+    use Psr\Container\ContainerInterface;
+
+    class CommandBus
+    {
+        /**
+         * @var ContainerInterface
+         */
+        private $handlerLocator;
+
+        // ...
+
+        public function handle(Command $command)
+        {
+            $commandClass = get_class($command);
+
+            if (!$this->handlerLocator->has($commandClass)) {
+                return;
+            }
+
+            $handler = $this->handlerLocator->get($commandClass);
+
+            return $handler->handle($command);
+        }
+    }
+
+The injected service is an instance of :class:`Symfony\\Component\\DependencyInjection\\ServiceLocator`
+which implements the PSR-11 ``ContainerInterface``, but it is also a callable::
+
+    // ...
+    $locateHandler = $this->handlerLocator;
+    $handler = $locateHandler($commandClass);
```

Comment:
I guess you mean `$this->handlerLocator($commandClass)` :). 
Anyway, the intermediate variable is mandatory, calling `$this->handlerLocator()` would make the engine look for a `handlerLocator()` method instead of executing the property as a callable.